src/jit/optimizer.cpp

```diff
@@ -2834,365 +2829,373 @@ void Compiler::optUnrollLoops()
-            test = test->gtStmt.gtStmtExpr;
-            noway_assert(incr->gtOper == GT_STMT);
-            incr = incr->gtStmt.gtStmtExpr;
+            /* Visit all the statements in the block */
 
-            // Don't unroll loops we don't understand.
-            if (incr->gtOper == GT_ASG)
+            for (GenTreeStmt* stmt = block->firstStmt(); stmt; stmt = stmt->gtNextStmt)
             {
-                continue;
-            }
+                /* Calculate gtCostSz */
+                gtSetStmtInfo(stmt);
 
-            /* Make sure everything looks ok */
-            if ((init->gtOper != GT_ASG) || (init->gtOp.gtOp1->gtOper != GT_LCL_VAR) ||
-                (init->gtOp.gtOp1->gtLclVarCommon.gtLclNum != lvar) || (init->gtOp.gtOp2->gtOper != GT_CNS_INT) ||
-                (init->gtOp.gtOp2->gtIntCon.gtIconVal != lbeg) ||
-
-                !((incr->gtOper == GT_ASG_ADD) || (incr->gtOper == GT_ASG_SUB)) ||
-                (incr->gtOp.gtOp1->gtOper != GT_LCL_VAR) || (incr->gtOp.gtOp1->gtLclVarCommon.gtLclNum != lvar) ||
-                (incr->gtOp.gtOp2->gtOper != GT_CNS_INT) || (incr->gtOp.gtOp2->gtIntCon.gtIconVal != iterInc) ||
-
-                (test->gtOper != GT_JTRUE))
-            {
-                noway_assert(!"Bad precondition in Compiler::optUnrollLoops()");
-                continue;
+                /* Update loopCostSz */
+                loopCostSz += stmt->gtCostSz;
             }
 
-            /* heuristic - Estimated cost in code size of the unrolled loop */
-
-            loopCostSz = 0;
-
-            block = head;
-
-            do
+            if (block == bottom)
             {
-                block = block->bbNext;
-
-                /* Visit all the statements in the block */
-
-                for (GenTreeStmt* stmt = block->firstStmt(); stmt; stmt = stmt->gtNextStmt)
-                {
-                    /* Get the expression and stop if end reached */
-
-                    GenTreePtr expr = stmt->gtStmtExpr;
-                    if (expr == incr)
-                    {
-                        break;
-                    }
-
-                    /* Calculate gtCostSz */
-                    gtSetStmtInfo(stmt);
-
-                    /* Update loopCostSz */
-                    loopCostSz += stmt->gtCostSz;
-                }
-            } while (block != bottom);
+                break;
+            }
+        }
 
-            /* Compute the estimated increase in code size for the unrolled loop */
+        /* Compute the estimated increase in code size for the unrolled loop */
 
-            unsigned int fixedLoopCostSz;
-            fixedLoopCostSz = 8;
+        unsigned int fixedLoopCostSz;
+        fixedLoopCostSz = 8;
 
-            int unrollCostSz;
-            unrollCostSz = (loopCostSz * totalIter) - (loopCostSz + fixedLoopCostSz);
+        int unrollCostSz;
+        unrollCostSz = (loopCostSz * totalIter) - (loopCostSz + fixedLoopCostSz);
```

Comment:
Do we know the first subexpression can't overflow here (realize it was like this before...)?